### PR TITLE
Add provider pre-fetching documentation

### PR DESCRIPTION
**Pull request by slontis**
Clearly document that implicit fetching is slower when using providers, and explain prefetching. Added to crypto.pod and migration_guide.pod links to it.

Add a link to EVP_default_properties_enable_fips() in crypto.pod.

##### Checklist

- [x] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

File: doc/man7/crypto.pod [hunk 1]
```diff
@@ -210,6 +211,60 @@ property string from the B<EVP_PKEY_CTX>.
...
+If you perform the same operation many times then it is recommeded to use
+L</Explicit fetching>. An algorithm may be prefetched once initially,
+and then passed to any operations that are currently using L</Implicit fetching>.
+See an example of Explicit fecthing in L</USING ALGORITHMS IN APPLICATIONS>.
+
+If the prefetched object is not passed to operations, then any implicit
+fetch will use the internally cached prefetched object, but it will
+still be slower than passing the prefetched object directly.
+
+Prior to OpenSSL 3.0, constant method tables (such as EVP_sha256()) were used
```
Comment (paulidale) [inline doc/man7/crypto.pod:225]: Something should be mentioned about these functions changing.  I.e. it's not clear that they aren't returning constant method tables in 3.x from the text here (sure, the implicit fetch above kind of says this, but making it clear in the performance section seems reasonable).

Comment (slontis) [reply doc/man7/crypto.pod:225]: Please suggest some text? I am not sure how that will make things clearer.

Comment (slontis) [reply doc/man7/crypto.pod]: I have attempted to update it, see if that helps..

**Review by paulidale — APPROVED** [review]

**Comment by slontis** [comment]
Added a bit more to the migration guide..

**Review by paulidale — APPROVED** [review]

File: doc/man7/crypto.pod [hunk 2]
```diff
@@ -210,6 +211,65 @@ property string from the B<EVP_PKEY_CTX>.
...
+
+If you perform the same operation many times then it is recommended to use
+L</Explicit fetching> to prefetch an algorithm once initially,
+and then pass this created object to any operations that are currently
+using L</Implicit fetching>.
+See an example of Explicit fetching in L</USING ALGORITHMS IN APPLICATIONS>.
+
+Prior to OpenSSL 3.0, constant method tables (such as EVP_sha256()) were used
+directly to access methods. If you pass one of these convenience functions
+to a operation the fixed methods are ignored, and only the name is used to
```
Comment (mattcaswell) [inline doc/man7/crypto.pod:224]: an operation

**Review by mattcaswell — CHANGES_REQUESTED** [review]

**Review by levitte — APPROVED** [review]
Pending fixing what @mattcaswell commented

**Comment by richsalz** [comment]
Shouldn't this also be documented, or linked, in the `EVP_sha256` etc man pages?
Yes, all of them.

**Comment by levitte** [comment]
> Shouldn't this also be documented, or linked, in the `EVP_sha256` etc man pages? Yes, all of them.

Linking would be the reasonable choice...  and yeah, a small NOTES section where the performance issue is mentioned, with a link to the documentation made here, that wouldn't be a bad idea.

**Comment by slontis** [comment]
Would this kind of thing be sufficient (before I go adding it all over the place)?

````
=head1 NOTES

Applications should be aware of the performance implications of
using this function multiple times and should consider using
L<EVP_MD_fetch(3)> instead.
See L<crypto(7)/Performance> for further information.
````

**Comment by paulidale** [comment]
Suggestions in bold:

> Applications should be aware of the **negative** performance implications of
> **calling** this function multiple times and should consider using
> L<EVP_MD_fetch(3)> instead.

**Comment by richsalz** [comment]
ChatGPT notwithstanding, maybe "Developers should be aware" as applications aren't sentient yet.

**Comment by slontis** [comment]
Thanks for the suggestions..

**Comment by slontis** [comment]
links added..

**Review by paulidale — APPROVED** [review]

**Review by mattcaswell — APPROVED** [review]

**Comment by mattcaswell** [comment]
Side note: we might want to make a similar documentation change to discuss SSL_CTX and how it is used to cache prefetched algorithms. Therefore you should avoid using a single SSL_CTX per connection.

File: doc/man7/crypto.pod [hunk 3]
```diff
@@ -213,21 +213,26 @@ property string from the B<EVP_PKEY_CTX>.
...
 fetch will use the internally cached prefetched object, but it will
 still be slower than passing the prefetched object directly.
 
-Prior to OpenSSL 3.0, constant method tables (such as EVP_sha256()) were used
-directly to access methods.
 Fetching via a provider offers more flexibility, but it is slower than the
 old method, since it must search for the algorithm in all loaded providers,
-and then build the method table. Internally OpenSSL caches similar algorithms
-on the first fetch (so loading a digest caches all digests).
+and then populate the method table using provider supplioed methods.
```
Comment (stefanberger) [inline doc/man7/crypto.pod]: s/supplioed/supplied

Comment (slontis) [reply doc/man7/crypto.pod]: Fixed in a later fixup commit..

**Comment by paulidale** [comment]
Merged.

**Pull request closed without merging**